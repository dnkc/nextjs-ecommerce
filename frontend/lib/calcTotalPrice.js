export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.product) return tally; //product deleted but still in your cart
    return tally + cartItem.quantity * cartItem.product.price;
  }, 0);
}
