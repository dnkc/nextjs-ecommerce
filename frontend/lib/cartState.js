import { createContext, useContext, useState } from "react";

const LocalStateContext = createContext();

const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // custom provider to store state and functionality - accessible via consumer
  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }
  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// custom hook for accessing cart local state
function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
