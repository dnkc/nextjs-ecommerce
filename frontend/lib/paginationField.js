import { PAGINATION_QUERY } from "../components/Pagination";

export default function paginationField() {
  return {
    keyArgs: false, //tells apollo that we will take care of everything
    read(existing = [], { args, cache }) {
      // args: first & skip values, cache is apollo cache
      const { skip, first } = args;
      // read number of items on the page from cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x); // filters if item is undefined

      // check if we do not have existing items

      // if there are items AND there are not enough items requested per page (default 2) AND we are on last page, then send it
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        // we do not have any items, go to network
        return false;
      }

      // if there are items, return them from the cache, do not go to network
      if (items.length) {
        console.log(
          `There are ${items.length} items in the cache! Gonna send them to apollo.`
        );
      }
      return items;
    },

    merge(existing, incoming, { args }) {
      // runs when apollo client comes back from the network with our product

      // when apollo tries to query for all products:
      // first thing: read function for those items
      // two things possible: return items bc they are in cache OR return false from here (makes a network request)
      console.log(`Merging items from the network ${incoming.length}`);
      const { skip, first } = args;
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      console.log(merged);
      // return merged items from cache
      return merged;
    },
  };
}
