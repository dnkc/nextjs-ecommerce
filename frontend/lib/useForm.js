import { useState, useEffect } from "react";

const useForm = (initialState = {}) => {
  // create a state object for our inputs

  const [inputs, setInputs] = useState(initialState);

  const initialValues = Object.values(initialState).join("");

  useEffect(() => {
    setInputs(initialState);
  }, [initialValues]);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === "number") {
      value = parseInt(value);
    }
    if (type === "file") {
      [value] = e.target.files;
    }
    setInputs({
      // copy the existing state
      ...inputs,
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initialState);
  }

  function clearForm() {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ""])
    );
    setInputs(blankState);
  }

  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
};

export default useForm;
