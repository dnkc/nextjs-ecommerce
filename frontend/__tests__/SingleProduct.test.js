import { MockedProvider } from "@apollo/react-testing";
import { render, screen } from "@testing-library/react";
import SingleProductPage, {
  SINGLE_ITEM_QUERY,
} from "../components/SingleProductPage";
import { fakeItem } from "../lib/testUtils";

const product = fakeItem();
const mocks = [
  {
    // When someone requests this query and variable combo
    request: {
      query: SINGLE_ITEM_QUERY,
      variables: {
        id: "123",
      },
    },
    // Return this data
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe("<SingleProduct />", () => {
  it("renders with proper data", async () => {
    // making fake data
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProductPage id="123" />
      </MockedProvider>
    );
    // debug();
    // wait for the test id to show up
    await screen.findByTestId("singleProduct");
    expect(container).toMatchSnapshot();
  });

  it("Errors out when an item is not found", async () => {
    const errorMocks = [
      {
        // When someone requests this query and variable combo
        request: {
          query: SINGLE_ITEM_QUERY,
          variables: {
            id: "123",
          },
        },
        // Return this data
        result: {
          errors: [{ message: "Item not found!!!!!!" }],
        },
      },
    ];
    const { container, debug } = render(
      <MockedProvider mocks={errorMocks}>
        <SingleProductPage id="123" />
      </MockedProvider>
    );
    // debug();
    // wait for the test id to show up
    await screen.findByTestId("graphql-error");
    expect(container).toHaveTextContent("Shoot!");
    expect(container).toHaveTextContent("Item not found!!!!!!");
  });
});
