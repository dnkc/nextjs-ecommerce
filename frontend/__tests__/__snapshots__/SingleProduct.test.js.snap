// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct /> renders with proper data 1`] = `
<div>
  <div
    class="SingleProductPage__ProductStyles-sc-10o3qp1-0 ieCAXR"
    data-testid="singleProduct"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
